google-cloud-clients/google-cloud-bigtable-admin/src/main/java/com/google/cloud/bigtable/admin/v2/BigtableTableAdminClient.java

```diff
@@ -327,18 +329,68 @@ public Table getTable(String tableId) {
    * }</pre>
    */
   public ApiFuture<List<TableName>> listTablesAsync() {
-    ApiFuture<ListTablesPagedResponse> listResp =
-        this.stub.listTablesPagedCallable().futureCall(composeListTableRequest());
+    ListTablesRequest request = ListTablesRequest.newBuilder().setParent(instanceName.toString())
+        .build();
 
-    return ApiFutures.transform(
-        listResp,
-        new ApiFunction<ListTablesPagedResponse, List<TableName>>() {
+    // TODO(igorbernstein2): try to upstream pagination spooling or figure out a way to expose the
+    // paginated responses while maintaining the wrapper facade.
+
+    // Fetch the first page.
+    ApiFuture<ListTablesPage> firstPageFuture = ApiFutures.transform(
+        stub.listTablesPagedCallable().futureCall(request),
+        new ApiFunction<ListTablesPagedResponse, ListTablesPage>() {
           @Override
-          public List<TableName> apply(ListTablesPagedResponse input) {
-            return convertToTableNames(input.iterateAll());
+          public ListTablesPage apply(ListTablesPagedResponse response) {
+            return response.getPage();
           }
         },
-        MoreExecutors.directExecutor());
+        MoreExecutors.directExecutor()
+    );
+
+    // Fetch the rest of the pages by chaining the futures.
+    ApiFuture<List<com.google.bigtable.admin.v2.Table>> allProtos = ApiFutures
+        .transformAsync(
+            firstPageFuture,
+            new ApiAsyncFunction<ListTablesPage, List<com.google.bigtable.admin.v2.Table>>() {
+              List<com.google.bigtable.admin.v2.Table> responseAccumulator = Lists
+                  .newArrayList();
+
+              @Override
+              public ApiFuture<List<com.google.bigtable.admin.v2.Table>> apply(
+                  ListTablesPage page) {
+                // Add all entries from the page
+                responseAccumulator.addAll(Lists.newArrayList(page.getValues()));
+
+                // If this is the last page, just return the accumulated responses.
+                if (!page.hasNextPage()) {
+                  return ApiFutures.immediateFuture(responseAccumulator);
+                }
+
+                // Otherwise fetch the next page.
+                return ApiFutures.transformAsync(
+                    page.getNextPageAsync(),
+                    this,
+                    MoreExecutors.directExecutor()
+                );
+              }
+            },
+            MoreExecutors.directExecutor()
+        );
+
+    // Wrap all of the accumulated protos.
+    return ApiFutures.transform(allProtos,
+        new ApiFunction<List<com.google.bigtable.admin.v2.Table>, List<TableName>>() {
+          @Override
+          public List<TableName> apply(List<com.google.bigtable.admin.v2.Table> protos) {
+            List<TableName> results = Lists.newArrayListWithCapacity(protos.size());
```

Comment:
nit: I prefer more specific naming than generic ones.

```
for (com.google.bigtable.admin.v2.Table table : tables) {
  names.add(TableName.parse(table.getName()));
}
```

Up to you.

Reply:
see above